src/modules/esl-carousel/renderers/esl-carousel.grid.renderer.ts

```diff
@@ -0,0 +1,96 @@
+  @memoize()
+  public get $fakeSlides(): HTMLElement[] {
+    const count = this.$carousel.$slides.length % this.ROWS;
+    if (count === 0) return [];
+    return Array.from({length: this.ROWS - count}, this.buildFakeSlide.bind(this));
+  }
+
+  /** @returns all slides including {@link ESLGridCarouselRenderer.$fakeSlides} slides created in grid mode */
+  public override get $slides(): HTMLElement[] {
+    return (this.$carousel.$slides || []).concat(this.$fakeSlides);
+  }
+
+  /** creates fake slide element */
+  public buildFakeSlide(): HTMLElement {
+    const $slide = document.createElement('div');
+    $slide.setAttribute('esl-carousel-fake-slide', '');
+    return $slide;
+  }
+
+  /**
+   * Processes binding of defined renderer to the carousel {@link ESLCarousel}.
+   * Prepare to renderer animation.
+   */
+  public override onBind(): void {
+    memoize.clear(this, '$fakeSlides');
+    this.$area.append(...this.$fakeSlides);
+    super.onBind();
+  }
+
+  /**
+   * Processes unbinding of defined renderer from the carousel {@link ESLCarousel}.
+   * Clear animation.
+   */
+  public override onUnbind(): void {
+    this.$fakeSlides.forEach((el) => el.remove());
+    super.onUnbind();
+  }
+
+  /**
+   * Processes changing slides
+   * Normalize actual active index to the first slide in the current dimension ('row')
+   */
+  public override async navigate(index: number, direction: ESLCarouselDirection, {activator}: ESLCarouselActionParams): Promise<void> {
+    await super.navigate(index - (index % this.ROWS), direction, {activator});
+  }
+
+  /** Processes animation. */
+  public override async onAnimate(nextIndex: number, direction: ESLCarouselDirection): Promise<void> {
+    const {activeIndex, $slidesArea} =  this.$carousel;
+    this.currentIndex = activeIndex;
+    if (!$slidesArea) return;
+    const step = this.ROWS * (direction === 'next' ? 1 : -1);
+    while (this.currentIndex !== nextIndex) await this.onStepAnimate(step);
+  }
+
+  protected override indexByOffset(offset: number): number {
+    return super.indexByOffset(offset * this.ROWS);
+  }
+
+  /**
+   * @returns count of slides to be rendered (reserved) before the first slide does not include fake slides
+   */
+  protected override calcReserveCount(back?: boolean): number {
+    const reserve = super.calcReserveCount(back);
+    return reserve - (reserve % this.ROWS);
+  }
+
+  /** Sets min size for slides */
+  protected override resize(): void {
+    if (!this.$area) return;
+    const areaStyles = getComputedStyle(this.$area);
+
+    this.gap = parseFloat(this.vertical ? areaStyles.rowGap : areaStyles.columnGap);
+    const areaSize = parseFloat(this.vertical ? areaStyles.height : areaStyles.width);
+    const count = Math.floor(this.count / this.ROWS);
+    this.slideSize = Math.floor((areaSize - this.gap * (count - 1)) / count);
+    this.$area.style.setProperty(ESLDefaultCarouselRenderer.SIZE_PROP, this.slideSize + 'px');
```

Comment:
I am not sure that SIZE_PROP is the proper name because I thought about things that weren't really there

Reply:
TBD for refactoring carousel. Thanks, I'll add it to list (and probably move list to the tickets).